infoleak/InfoLeak.java

```
@@ -0,0 +1,223 @@
+    void verifySignature(AccessibleServletRequest request) throws URIException {
+        try {
+            String sig = getSig(request);
+
+            Validate.notNull(sig, "'sig' is required!");
+            String serial = getSerial(request);
+            Validate.notNull(serial, "'serial' is required!");
+
+            String parsedUrl = getUrlToSign(request);
+            logger.info("{"
+                + "\n requestUrl:" + RedactUtils.redact(request.getRequestURL().toString() +
+                (request.getQueryString() == null ? "" : "?" + request.getQueryString()), sig)
+                + "\n parsedUrl:" + parsedUrl
+                + "\n MaterialSet:" + materialName
+                + "\n method:" + request.getMethod() + "}");
+
+            byte[] sigBytes = getSigBytes(sig);
+
+            byte[] data = org.apache.commons.codec.binary.StringUtils.getBytesUtf8(parsedUrl);
+
+            //Setting to false to override any user-supplied attribute
+            request.setAttribute(VALIDSIG_ATTR, false);
+            if(!RequestMethod.OPTIONS.name().equals(request.getMethod())) {
+                verifySignature(Long.parseLong(serial), data, sigBytes);
+            }
+            request.setAttribute(VALIDSIG_ATTR, true);
+        }
+        catch (IllegalArgumentException e) {
+            // Any IllegalArgument to the signature-handler must be an attempt to
+            // change the signature. 403, not 400.
+            throw new SignatureMismatchException(e);
+        }
+    }
+
+    protected void verifySignature(long serial, byte[] data, byte[] actualSig) {
+        // 0 means use the latest serial to OdinKeyPair and that is not valid
+        // behavior for CAPS.
+        if (serial == 0) {
+            throw new IllegalArgumentException("Invalid Serial");
+        }
+        OdinKeyPair keyPair = keys.computeIfAbsent(serial, ser -> {
+            try {
+                return new OdinKeyPair(materialName, ser);
+            } catch (Exception e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)